Mage/src/main/java/mage/abilities/keyword/CraftAbility.java

```diff
@@ -0,0 +1,147 @@
+package mage.abilities.keyword;
+
+import mage.abilities.Ability;
+import mage.abilities.ActivatedAbilityImpl;
+import mage.abilities.costs.Cost;
+import mage.abilities.costs.CostImpl;
+import mage.abilities.costs.common.ExileSourceCost;
+import mage.abilities.costs.mana.ManaCostsImpl;
+import mage.abilities.effects.OneShotEffect;
+import mage.cards.Card;
+import mage.constants.Outcome;
+import mage.constants.TimingRule;
+import mage.constants.Zone;
+import mage.filter.FilterCard;
+import mage.filter.FilterPermanent;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.common.TargetCardInGraveyardBattlefieldOrStack;
+import mage.util.CardUtil;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ * @author TheElk801
+ */
+public class CraftAbility extends ActivatedAbilityImpl {
+
+    private final String description;
+    private final String manaString;
+
+    public CraftAbility(String manaString, String description, FilterCard filterCard, FilterPermanent filterPermanent) {
```

Comment:
Should have a comment reminding that the owner predicate is needed on the card filter, anotherpredicate is needed on the permanent filter, and any other usage notes that might be helpful

actually... is it possible to have some sort of verify check for the presence of those predicates?

okay third idea... how about instead of taking a FilterCard and a FilterPermanent, we take a list of Predicate<MageObject> and construct the proper filters? Could also have an optional param for number of targets if all the same.

Reply:
yeah I might do something like that actually